### PR TITLE
feat: make armadillo work with local storage instead of minio

### PR DESCRIPTION
**Pull request by mswertz**
Needs a rebase after #196 and #197 are merged

TODO:
- [x] Put MinIO behind a docker-compose profile so it is not automatically started when developing
- [x] Manual test with http examples
- [x] Manual test with R script examples in `docker/test`
- [x] Fix unit tests
- [x] Update descriptions of example files where data is loaded
- [x] Resolve SonarCloud issues

### CODE REVIEW

**Comment by mswertz** [comment]
[Implement API to manage profiles](https://trello.com/c/hjiXHQcg/918-implement-api-to-manage-profiles)

**Comment by mswertz** [comment]
[Use local storage](https://trello.com/c/duN6he4b/919-use-local-storage)

File: armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  @Override
+  public boolean objectExists(String bucket, String objectName) {
+    Objects.requireNonNull(bucket);
+    Objects.requireNonNull(objectName);
+
+    // no uppercase in bucket name
+    if (bucket.matches(".*[A-Z].*")) {
+      // unsure why this matters
```
Comment (tommydeboer) [inline armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java:29]: Bucket naming rules:
https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html

Comment (tommydeboer) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java:29]: For local storage we can remove this rule, unless you want to keep it consistent between MinIO and local

Comment (tommydeboer) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java:29]: Keeping it consistent seems a smart move for now, I would say. Then we can migrate back and forth if needed without any problems

File: armadillo/src/main/java/org/molgenis/armadillo/storage/ObjectMetadata.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package org.molgenis.armadillo.storage;
+
+import io.minio.messages.Item;
+import java.io.File;
+import java.util.Date;
+
+public class ObjectMetadata {
```
Comment (tommydeboer) [inline armadillo/src/main/java/org/molgenis/armadillo/storage/ObjectMetadata.java:7]: Let's use a  `record` here since the class is immutable

File: armadillo/src/main/resources/application.yml [hunk 3]
```diff
@@ -14,11 +14,11 @@ datashield:
         datashield:
           # the seed can only be 9 digits
           seed: 342325352
-         # foo: bar
+          # foo: bar
 
 springdoc:
   swagger-ui:
-    enabled: false
+    enabled: true
```
Comment (tommydeboer) [inline armadillo/src/main/resources/application.yml:21]: Move to separate commit and/or PR

File: armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java (hunk 1 above)
Comment (mswertz) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java:29]: agreed

Comment (mswertz) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java:29]: might be good to centralize the regexp so we share them between implementations?

File: armadillo/src/main/java/org/molgenis/armadillo/storage/ObjectMetadata.java (hunk 2 above)
Comment (mswertz) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/ObjectMetadata.java:7]: indeed.

File: armadillo/src/main/resources/application.yml (hunk 3 above)
Comment (mswertz) [reply armadillo/src/main/resources/application.yml:21]: agreed

File: armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java (hunk 1 above)
Comment (tommydeboer) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/LocalStorageService.java:29]: Good idea

File: armadillo/src/test/java/org/molgenis/armadillo/storage/LocalStorageServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  LocalStorageService localStorageService;
+
+  @BeforeEach
+  void beforeEach() throws IOException {
+    String tmpDir = Files.createTempDirectory("armadilloStorageTest").toFile().getAbsolutePath();
+    localStorageService = new LocalStorageService(tmpDir);
+  }
+
+  @Test
+  void testCheckBucketExistsCreatesBucketIfNotFound() {
```
Comment (mswertz) [inline armadillo/src/test/java/org/molgenis/armadillo/storage/LocalStorageServiceTest.java:29]: minor: replace Bucket with Project

**Review by mswertz — COMMENTED** [review]
I only have minor comments inline, code review looks good. 

It would probably be good to update README so we can explain for developers what needs to be done to either run on minio or run with local storage?

File: .gitignore [hunk 5]
```diff
@@ -1,3 +1,5 @@
+data/user-*
```
[the remark this answers is not in the document]
Comment (mswertz) [reply .gitignore:1]: ah, now I understand, you want to get rid of user data and keep the examples. Good.

File: armadillo/src/main/java/org/molgenis/armadillo/storage/ArmadilloStorageService.java [hunk 6]
```diff
@@ -21,33 +19,32 @@
...
   @PostFilter("hasAnyRole('ROLE_SU', 'ROLE_' + filterObject.toUpperCase() + '_RESEARCHER')")
+  @SuppressWarnings("java:S6204") // result of method can't be unmodifiable because of @PostFilter
   public List<String> listProjects() {
-    return storageService.listBuckets().stream()
-        .map(Bucket::name)
+    return storageService.listProjects().stream()
         .filter(it -> it.startsWith(SHARED_PREFIX))
         .map(it -> it.substring(SHARED_PREFIX.length()))
-        .collect(toList());
+        .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (tommydeboer) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/ArmadilloStorageService.java:37]: In this single case we actually can't! 😄 See `@SuppressWarning` above the method

[the remark this answers is not in the document]
Comment (tommydeboer) [reply armadillo/src/main/java/org/molgenis/armadillo/storage/ArmadilloStorageService.java:37]: Ah, but you mean the static import probably. Changing 👍

**Review by tommydeboer — APPROVED** [review]
